docs/core/middleware.md

```diff
@@ -0,0 +1,158 @@
+<!--
+order: 3
+-->
+
+# Middlewares
+
+
+`Middlewares` are objects that the developer can create to add logic before or after the transaction handler in a composable manner.
+{synopsis}
+
+## Pre-requisite Readings
+
+* [Anatomy of a Cosmos SDK Application](../basics/app-anatomy.md) {prereq}
+* [Transactons](transactions.md){prereq}
+
+
+## Middlewares
+
+The SDK Baseapp's implementation of ABCI CheckTx, DeliverTx, and Baseapp's own Simulate methods use a middleware-based design. Middlewares can add logic to be executed before or after a transaction handler execution. Middlewares are like an `antehandler` with the added feature of being able to add post-transaction handler execution. Middlewares allow us to solve use cases like transaction Tips and refund unused gas(issue [#2150](https://github.com/cosmos/cosmos-sdk/issues/2150)). 
+
+
+### Type Definition
+
+The two following interfaces are the base of the middleware design, and are defined in types/tx:
+
++++ https://github.com/cosmos/cosmos-sdk/blob/5491be27d02e796746bd78d3d08bd1b2a9b1deb2/types/tx/middleware.go#L62-L71
+
+Where we define the following arguments and return types:
+
++++ https://github.com/cosmos/cosmos-sdk/blob/5491be27d02e796746bd78d3d08bd1b2a9b1deb2/types/tx/middleware.go#L25-L60
+
+BaseApp holds a reference to a `tx.Handler`:
```

Comment:
```suggestion
BaseApp holds a reference to a `tx.Handler`, which will process the incoming transactions. It could be a simple `sdk.Msg` executor, but in practice, it will be defined as a middleware stack wrapping around the base `sdk.Msg` executor.
```